File: exercise/scala/day24/src/main/scala/domain/Toy.scala
package domain

import domain.core.{Error, EventSourcedAggregate}

import java.time.LocalDateTime
import java.util.UUID
import scala.util.{Either, Left, Right}

class Toy private(
                   timeProvider: () => LocalDateTime,
                   private var name: String,
                   private var stock: StockUnit
                 ) extends EventSourcedAggregate(timeProvider) {
  raiseEvent(ToyCreatedEvent(UUID.randomUUID(), timeProvider(), name, stock))

  def getName(): String = name

  def reduceStock(): Either[Error, Toy] = {
    if (!stock.isSupplied) return Left(Error(s"No more $name in stock"))
    raiseEvent(StockReducedEvent(id, time(), name, stock.decrease))
    Right(this)
  }

  override def registerRoutes(): Unit = {
    registerEventRoute(classOf[ToyCreatedEvent], this.apply)
    registerEventRoute(classOf[StockReducedEvent], this.apply)
  }

  private def apply(event: ToyCreatedEvent): Unit = {
    _id = event.id
    name = event.name
    stock = event.stock
  }

  private def apply(event: StockReducedEvent): Unit = stock = event.newStock
}

object Toy {
  def create(timeProvider: () => LocalDateTime, name: String, stock: Int): Either[Error, Toy] =
    StockUnit.from(stock).map(stockUnit => new Toy(timeProvider, name, stockUnit))
}

File: exercise/java/day24/src/main/java/domain/Toy.java
package domain;

import domain.core.EventSourcedAggregate;
import io.vavr.control.Either;
import lombok.Getter;

import java.time.LocalDateTime;
import java.util.UUID;
import java.util.function.Supplier;

import static io.vavr.control.Either.left;
import static io.vavr.control.Either.right;

public class Toy extends EventSourcedAggregate {
    @Getter
    private String name;
    private StockUnit stock;

    private Toy(Supplier<LocalDateTime> timeProvider, String name, StockUnit stock) {
        super(timeProvider);
        raiseEvent(new ToyCreatedEvent(UUID.randomUUID(), timeProvider.get(), name, stock));
    }

    public static Either<Error, Toy> create(Supplier<LocalDateTime> timeProvider, String name, int stock) {
        return StockUnit
                .from(stock)
                .map(stockUnit -> new Toy(timeProvider, name, stockUnit));
    }

    private void apply(ToyCreatedEvent event) {
        setId(event.id());
        name = event.name();
        stock = event.stock();
    }

    public Either<domain.core.Error, Toy> reduceStock() {
        if (!stock.isSupplied()) return left(new domain.core.Error("No more " + name + " in stock"));
        raiseEvent(new StockReducedEvent(getId(), time(), name, stock.decrease()));

        return right(this);
    }

    private void apply(StockReducedEvent event) {
        stock = event.newStock();
    }

    @Override
    protected void registerRoutes() {
        registerEventRoute(ToyCreatedEvent.class, this::apply);
        registerEventRoute(StockReducedEvent.class, this::apply);
    }
}